TickSpec/FeatureParser.fs

```diff
@@ -1,92 +1,131 @@
 ﻿module TickSpec.FeatureParser
 
 open System.Text.RegularExpressions
-open TickSpec.LineParser
 open TickSpec.BlockParser
+open System
 
-/// Computes combinations of table values
-let internal computeCombinations (tables:Table []) =
-    let values =
-        tables
-        |> Seq.map (fun table ->
-            table.Rows |> Array.map (fun row ->
-                row
-                |> Array.mapi (fun i col ->
-                    table.Header.[i],col
-                )
-            )
-        )
-        |> Seq.toList
-    values |> List.combinations
-
-/// Replace line with specified named values
-let internal replaceLine (xs:seq<string * string>) (scenario,n,tags,line,step) =
-    let replace s =
+/// Parses lines of feature
+let parseFeature (lines:string[]) =
+    let replace combination s =
         let lookup (m:Match) =
             let x = m.Value.TrimStart([|'<'|]).TrimEnd([|'>'|])
-            xs |> Seq.tryFind (fun (k,_) -> k = x)
+            combination |> Seq.tryFind (fun (k,_) -> k = x)
             |> (function Some(_,v) -> v | None -> m.Value)
         let pattern = "<([^<]*)>"
         Regex.Replace(s, pattern, lookup)
-    let step =
-        match step with
-        | GivenStep s -> replace s |> GivenStep
-        | WhenStep s -> replace s |> WhenStep
-        | ThenStep s  -> replace s |> ThenStep
-    let table =
-        line.Table
-        |> Option.map (fun table ->
-            Table(table.Header,
-                table.Rows |> Array.map (fun row ->
-                    row |> Array.map (fun col -> replace col)
-                )
+
+    let computeCombinations examples =
+        let rec combinations source =
+            match source with
+            | [] -> [[]]
+            | (header, rows) :: xs ->
+                [ for row in rows do
+                    for combinedRow in combinations xs ->
+                        (header, row) :: combinedRow ]
+
+        let processRow rowSet =
+            rowSet
+            |> List.fold (fun state (_, (rowMap, rowTags)) ->
+                match state with
+                | None -> None
+                | Some (stateTags, stateMap) ->
+                    let newStateMap =
+                        rowMap
+                        |> Map.fold (fun state key value ->
+                            match state with
+                            | None -> None
+                            | Some map ->
+                                let existingValue = map |> Map.tryFind key
+                                match existingValue with
+                                | None -> Some (map.Add (key, value))
+                                | Some x when x = value -> Some map
+                                | _ -> None) (Some stateMap)
```

Comment:
when doing a fold, as per the F# coding guidelines, it can work well to do

`(Some stateMap,rowMap) ||> Map.fold (fun ...` so the correlated bits sit together and the initial state doesnt ahve to hand at the end

Reply:
Thanks, changed.